cpp/src/parquet/column_reader.cc

```diff
@@ -1329,6 +1332,185 @@ class TypedRecordReader : public ColumnReaderImplBase<DType>,
+         // If we had exhausted the last page, HasNextInternal() must have advanced
+         // to the next page. So there must be available values to process.
+         if (batch_size == 0) {
+        break;
+         }
+
+         // For skip we will read the levels and append them to the end
+         // of the def_levels and rep_levels just like for read.
+         ReserveLevels(batch_size);
+
+         int16_t* def_levels = this->def_levels() + levels_written_;
+         int16_t* rep_levels = this->rep_levels() + levels_written_;
+
+         int64_t levels_read = 0;
+         levels_read = this->ReadDefinitionLevels(batch_size, def_levels);
+         if (this->ReadRepetitionLevels(batch_size, rep_levels) != levels_read) {
+        throw ParquetException("Number of decoded rep / def levels did not match");
+         }
+
+         levels_written_ += levels_read;
+
+         // Look at the buffered levels, delimit them based on
+         // (rep_level == 0), report back how many records are in there, and
+         // fill in how many not-null values (def_level == max_def_level_).
+         // DelimitRecords updates levels_position_.
+         int64_t start_levels_position = levels_position_;
+         int64_t remaining_records = num_records - skipped_records;
+         int64_t values_seen = 0;
+         skipped_records += DelimitRecords(remaining_records, &values_seen);
+         if (ReadAndThrowAway(values_seen) != values_seen) {
+        throw ParquetException("Could not read and throw away requested values");
+         }
+         // Mark those levels and values as consumed in the the underlying page.
+         // This must be done before we throw away levels since it updates
+         // levels_position_ and levels_written_.
+         this->ConsumeBufferedValues(levels_position_ - start_levels_position);
+         // Updated levels_position_ and levels_written_.
+         ThrowAwayLevels(start_levels_position);
+    }
+
+    return skipped_records;
+  }
+
+  // Read 'num_values' values and throw them away.
+  int64_t ReadAndThrowAway(int64_t num_values) {
+    int64_t values_left = num_values;
+    int64_t batch_size = kMinLevelBatchSize;  // ReadBatch with a smaller memory footprint
+    int64_t values_read = 0;
+
+    // This will be enough scratch space to accommodate 16-bit levels or any
+    // value type
+    int value_size = type_traits<DType::type_num>::value_byte_size;
+    std::shared_ptr<ResizableBuffer> scratch = AllocateBuffer(
+        this->pool_, batch_size * std::max<int>(sizeof(int16_t), value_size));
+    do {
+      batch_size = std::min<int>(batch_size, values_left);
+      values_read = this->ReadValues(
+          batch_size, reinterpret_cast<T*>(scratch->mutable_data()));
+      values_left -= values_read;
+    } while (values_read > 0 && values_left > 0);
+    return num_values - values_left;
+  }
+
+  int64_t SkipRecords(int64_t num_records) override {
```

Comment:
The idea is to enable Skipping entire pages at a time if the number of records to skip is sufficiently large.

Reply:
Can you explain where the "Skipping entire pages at a time" part happens?

Reply:
To give some context, suppose that we have M records in the column chunk and we want to read only record N from the column. In that case, we can do something like this. SkipRecords(N-1), ReadRecords(1), SkipRecords(M - N). We can do some optimizations here. For example for the skip to the end of the column chunk, we would not need to even look at page headers.

Reply:
Ok, but are those optimizations done in this PR?

Reply:
No, here is what we have now: Consider a non-repeated field, and that there are 10 pages with 100 values each. SkipRecords(900) will skip "decoding" the first 9 pages. It will still look at the page headers to find out how many values there are per page.

Reply:
It happens in TypedColumnReader::Skip.